docs/source/compilation-databases.rst

```diff
@@ -0,0 +1,126 @@
+
+    cmake_minimum_required(VERSION 3.5)
+    project(tutorial)
+
+    set(CMAKE_EXPORT_COMPILE_COMMANDS ON)
+
+    set(SOURCES main.cpp third-party/library.cpp)
+
+    option(GPU_OFFLOAD "Enable GPU offload." OFF)
+    if (GPU_OFFLOAD)
+        add_definitions("-D GPU_OFFLOAD=1")
+        list(APPEND SOURCES gpu/foo.cpp)
+    else()
+        list(APPEND SOURCES cpu/foo.cpp)
+    endif()
+
+    add_executable(tutorial ${SOURCES})
+
+.. important::
+    For projects that don't use CMake, we can use `Bear`_ to intercept the
+    commands generated by other build systems (such as GNU makefiles). Other
+    build systems and tools that produce compilation databases should also be
+    compatible.
+
+.. _`compilation database`: https://clang.llvm.org/docs/JSONCompilationDatabase.html
+.. _`Bear`: https://github.com/rizsotto/Bear
+
+
+CPU Compilation Commands
+------------------------
+
+Let's start by running CMake without the :code:`GPU_OFFLOAD` option enabled, to
+obtain a compilation database for the CPU:
+
+.. code :: sh
+
+    $ mkdir build-cpu
+    $ cmake ../
+    $ ls
+
+    CMakeCache.txt  CMakeFiles  Makefile  cmake_install.cmake  compile_commands.json
+
+This :code:`compile_commands.json` file includes all the commands required to
+build the code, corresponding to the commands that would be executed if we were
+to actually run :code:`make`.
+
+.. attention::
+    CMake generates compilation databases when the ``cmake`` command is
+    executed, allowing us to generate compilation databases without also
+    building the application. Other tools (like Bear) may require a build.
+
+In this case, it contains::
+
+    [
+    {
+      "directory": "/home/username/src/build-cpu",
+      "command": "/usr/bin/c++ -o CMakeFiles/tutorial.dir/main.cpp.o -c /home/username/src/main.cpp",
+      "file": "/home/username/src/main.cpp"
+    },
+    {
+      "directory": "/home/username/src/build-cpu",
+      "command": "/usr/bin/c++ -o CMakeFiles/tutorial.dir/third-party/library.cpp.o -c /home/username/src/third-party/library.cpp",
+      "file": "/home/username/src/third-party/library.cpp"
+    },
+    {
+      "directory": "/home/username/src/build-cpu",
+      "command": "/usr/bin/c++ -o CMakeFiles/tutorial.dir/cpu/foo.cpp.o -c /home/username/src/cpu/foo.cpp",
+      "file": "/home/username/src/cpu/foo.cpp"
+    }
+    ]
+
+
+GPU Compilation Commands
+------------------------
+
+Repeating the exercise with :code:`GPU_OFFLOAD` enabled gives us a different
+compilation database for the GPU.
+
+.. warning::
+    The ``GPU_OFFLOAD`` option is specific to this ``CMakeLists.txt`` file, and
+    isn't something provided by CMake. Understanding how to build an application
+    for a specific target platform is beyond the scope of this tutorial.
+
+As expected, we can see that the compilation database refers to ``gpu.cpp``
+instead of ``cpu.cpp``, and that the ``GPU_OFFLOAD`` macro is defined as part
+of each compilation command::
+
+    [
```

Comment:
Meant to be in a code block or does it render correctly?

Reply:
Fixed in 123d24efd632d2f5155281c2b8c6a4715e3c12ad.